pkg/speaker/manager.go

```diff
@@ -0,0 +1,415 @@
+package speaker
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/openelb/openelb/api/v1alpha2"
+	"github.com/openelb/openelb/pkg/constant"
+	"github.com/openelb/openelb/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type handleResult int
+
+const (
+	none handleResult = iota
+	wantDelete
+	wantStore
+	wantReset
+)
+
+type recordInfo struct {
+	ips       []string
+	eip       string
+	protocols string
+	speaker   string
+}
+
+func (r *recordInfo) String() string {
+	return fmt.Sprintf("{eip:%s, protocol:%s, speaker:%s, ips:%s}",
+		r.eip, r.protocols, r.speaker, strings.Join(r.ips, ";"))
+}
+
+type speaker struct {
+	Speaker
+	ch chan struct{}
+}
+
+type Manager struct {
+	sync.RWMutex
+	client.Client
+	logr.Logger
+
+	// map[name]Speaker
+	speakers map[string]speaker
+
+	// map[eip.name]EIP
+	eips map[string]*v1alpha2.Eip
+
+	// map[NamespacedName]recordInfo
+	ips map[string]*recordInfo
+}
+
+func NewSpeakerManager(c client.Client, l logr.Logger) *Manager {
+	return &Manager{
+		Client:   c,
+		Logger:   l,
+		speakers: make(map[string]speaker, 0),
+		eips:     make(map[string]*v1alpha2.Eip, 0),
+		ips:      make(map[string]*recordInfo, 0),
+	}
+}
+
+func (m *Manager) RegisterSpeaker(name string, s Speaker) error {
+	t := speaker{
+		Speaker: s,
+		ch:      make(chan struct{}),
+	}
+
+	err := s.Start(t.ch)
+	if err == nil {
+		m.speakers[name] = t
+		return nil
+	}
+
+	return err
+}
+
+func (m *Manager) UnRegisterSpeaker(name string) {
+	t, ok := m.speakers[name]
+	if ok {
+		close(t.ch)
+	}
+	delete(m.speakers, name)
+}
+
+func (m *Manager) GetSpeaker(name string) Speaker {
+	t, ok := m.speakers[name]
+	if ok {
+		return t.Speaker
+	}
+
+	return nil
+}
+
+func (m *Manager) HandleEIP(eip *v1alpha2.Eip) error {
+	if eip == nil {
+		return nil
+	}
+
+	m.Lock()
+	defer m.Unlock()
+	_, exist := m.eips[eip.GetName()]
+	if exist {
+		if !eip.DeletionTimestamp.IsZero() {
+			m.V(3).Info(fmt.Sprintf("deleting eip:[%s]", eip.GetName()))
+			delete(m.eips, eip.GetName())
+		}
+
+		return nil
+	}
+
+	if !eip.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	// TODO: layer2 validate NIC infos
+	m.V(3).Info(fmt.Sprintf("store eip:[%s]", eip.GetName()))
+	m.eips[eip.GetName()] = eip
+	return nil
+}
+
+func (m *Manager) HandleService(svc *corev1.Service) error {
+	if svc == nil {
+		return nil
+	}
+
+	svcNSName := types.NamespacedName{Namespace: svc.GetNamespace(), Name: svc.GetName()}.String()
+
+	// get local record info, if localRecord is nil -- no record exists(initing)
+	localRecord, exist := m.ips[svcNSName]
+	if !exist && svc.Spec.Type != corev1.ServiceTypeLoadBalancer {
```

Comment:
If there is a case where one `exist` equals false and the type of `svc` is lb, how will it be handled?

Reply:
> If there is a case where one `exist` equals false and the type of `svc` is lb, how will it be handled?

We will continue to process downwards, obtain the relevant records of the service, store them in `m.ips` and announce the route.